Allow connection consumers to force a reconnect

leesio opened the pull request:
Consumers of the library can set timeouts on operations and force a
reconnect in the event of timeouts. This moves bad connections more
quickly into a disconnected state, allowing consumers to remove them
from the connection pool into a new connection has been successfully
established. For example, in the [jobsworth](https://github.com/pusher/jobsworth):

```
            c.put(@marshaller.dump(job), options).timeout(@timeout).errback {
              @logger.warn("put timed out to #{c.host}")
              if c.connected?
                @logger.info("forcing reconnect to #{c.host}")
                c.force_reconnect
              end
            }
```

There is an argument that the semantics of `intentionally_closed`
don't quite fit with this use case - in the force_reconnect method, the
connection was `intentionally_closed` too - a better description might
be `desired_state = connected` or `should_be_connected`. Open to
suggestions

Commit:
Allow connection consumers to force a reconnect

Consumers of the library can set timeouts on operations and force a
reconnect in the event of timeouts. This moves bad connections more
quickly into a disconnected state, allowing consumers to remove them
from the connection pool into a new connection has been successfully
established

There is an argument that the semantics of `intentionally_closed`
don't quite fit with this use case - in the force_reconnect method, the
connection was `intentionally_closed` too - a better description might
be `desired_state = connected` or `should_be_connected`. Open to
suggestions

WillSewell commented:
What about renaming it `should_reconnect`?

leesio commented:
I did consider that, but there's another case that's covered by the `desired_state` thing:

If `disconnect` is called before the `connected` handler fires, it sets `intentionally_closed` to true, this indicates to the `connected` handler that, although the connection has succeeded, the `desired_state` is disconnected, so it closes the connection.

I think the semantics of `should_reconnect` don't work here.

i.e.
```
def connected

  ...
  unless @should_reconnect
    @conn.close_connection_after_writing
  end
end
```

We could use another variable, I suppose, but I think this `intentionally_closed`/`desired_state` thing encapsulates what's required quite nicely.

leesio commented:
>  I'll let you decide what to call it
Thanks! 😅

mdpye commented on lib/em-jack/connection.rb:
```diff
@@ -284,6 +284,11 @@ def disconnect
     end
 
+    def force_reconnect
+      @intentionally_closed = false
```
Is there a reason (other than an abundance of caution) that you set `@intentionally_closed` before forcing a reconnect? It can only be `true` if `disconnect` has been called in the past.

I see a few other suggestions in the code which suggest that no operations are valid on the `EMJack::Connection` after `disconnect` has been called. For example `reconnect!` doesn't clean the state of `intentionally_disconnected` either. Is it valid to call `disconnect` and then later `reconnect!`? If it is, then it's likely buggy and should be fixed as well.

But I suspect that it isn't, and so calling `force_reconnect` after `disconnect` should probably leave the value of `@intentionally_closed` alone.

leesio replied:
I think your initial assumption of abundant caution is accurate. I'll rectify.

mdpye commented on lib/em-jack/connection.rb:
```diff
@@ -284,6 +284,11 @@ def disconnect
       @close_df = EM::DefaultDeferrable.new
     end
 
+    def force_reconnect
```
Confusingly, by Ruby conventions, `force_reconnect` and `reconnect!` read as more or less the same name. What is `reconnect!` for, and if it doesn't serve your purpose, is that cos it's buggy rather than because it wasn't meant to?

leesio replied:
I was aware there was overlap there, I just wasn't sure whether anything else depended on the behaviour of that `reconnect!` method. I suspect `force_reconnect` may be a better alternative though so should probably just replace it and see if anything breaks 🙈

Commit:
Use the existing reconnect! method to reconnect

- use a @desired_state variable rather than intentionally_closed - it
better reflected the semantics here
- don't change @desired state on reconnect!

WillSewell commented on lib/em-jack/connection.rb:
```diff
@@ -322,13 +324,10 @@ def disconnected
-      EM.next_tick do
-        @conn.reconnect(@host, @port)
-        initialize_tube_state
-      end
```
Do you know what the lines you removed do? Have you checked whether we use this method for anything?

leesio replied:
The previous `reconnect!` method:

- reset the retries count (this happens in the `connected` handler, which I think is right)
- resets the tube state (this happens in the `disconnected` handler, which will be called when the disconnection which we trigger occurs)
- Calls `reconnect` on the `@conn` which will do nothing if the state is still `connected`
- initialises the tube state (this also happens in the `disconnected` handler

I don't think any of the stuff that happens in the old method won't happen with the new method!

As for whether anything uses it - I can't see anything from `grep`-ing in pusher-server and jobsworth!

It's also been running on `eu` overnight and everything seems fine!
